{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "ef31330f",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "from tqdm import tqdm\n",
    "import argparse"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "annotation_path = \"/scratch/tsoies-DL/Output_DNBelab/Acomys_dimidiatus.gtf\"\n",
    "annotation = pd.read_csv(annotation_path, sep='\\t', names=['scaffold', 'produced_by', 'feature_type', 'start', 'end', 'dot', 'strand', 'frame', 'desc'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ca2583c6",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "from tqdm import tqdm\n",
    "import argparse\n",
    "class process_annotation():\n",
    "    def __init__(self, annotation_path, add_type=True):\n",
    "        self.annotation_path = annotation_path\n",
    "        self.annotation = pd.read_csv(self.annotation_path, sep='\\t', names=['scaffold', 'produced_by', 'feature_type', 'start', 'end', 'dot', 'strand', 'frame', 'desc'])\n",
    "        self.add_type= add_type\n",
    "    def desc2fullDesc(self):\n",
    "        desc2fullDescDict = {}\n",
    "        unique_desc_full = annotation.desc[self.annotation.desc.apply(lambda x: '_id' in x)].unique()\n",
    "        for desc in tqdm(self.annotation.desc):\n",
    "            if \"_id\" in desc:\n",
    "                desc2fullDescDict[desc] = desc\n",
    "            else:\n",
    "                for full_desc in unique_desc_full:\n",
    "                    if f'\"{desc}\"' in full_desc:\n",
    "                        desc2fullDescDict[desc] = full_desc\n",
    "        return desc2fullDescDict\n",
    "    def process_description(self):\n",
    "        desc2fullDescDict = self.desc2fullDesc()\n",
    "        annotation = self.annotation\n",
    "        annotation['desc'] = annotation['desc'].apply(lambda x: desc2fullDescDict[x])\n",
    "        if self.add_type: annotation['desc'] = annotation['desc'] + '; gene_biotype \"protein_coding\";'\n",
    "        return annotation\n",
    "    \n",
    "if __name__ == 'main':\n",
    "    parser = argparse.ArgumentParser(prog='process_annotation')\n",
    "    parser.add_argument('-i', 'ingtf')\n",
    "    parser.add_argument('-o', 'outgtf')\n",
    "    args = parser.parse_args()\n",
    "    outgtf = process_annotation(args.ingtf).process_description()\n",
    "    outgtf.to_csv(args.outgtf, index=False, sep='\\t', header=False)\n",
    "\n",
    "    \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "37d66b23",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "  0%|          | 1056/332336 [00:00<01:01, 5380.80it/s]"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 332336/332336 [00:56<00:00, 5831.52it/s]\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>scaffold</th>\n",
       "      <th>produced_by</th>\n",
       "      <th>feature_type</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>dot</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>desc</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CAJQZA010000166.1</td>\n",
       "      <td>GeneMark.hmm3</td>\n",
       "      <td>gene</td>\n",
       "      <td>110</td>\n",
       "      <td>817</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript_id \"g1.t1\"; gene_id \"g1\";; gene_bio...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CAJQZA010000166.1</td>\n",
       "      <td>GeneMark.hmm3</td>\n",
       "      <td>transcript</td>\n",
       "      <td>110</td>\n",
       "      <td>817</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript_id \"g1.t1\"; gene_id \"g1\";; gene_bio...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CAJQZA010000166.1</td>\n",
       "      <td>GeneMark.hmm3</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>110</td>\n",
       "      <td>112</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>transcript_id \"g1.t1\"; gene_id \"g1\";; gene_bio...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CAJQZA010000166.1</td>\n",
       "      <td>GeneMark.hmm3</td>\n",
       "      <td>CDS</td>\n",
       "      <td>110</td>\n",
       "      <td>160</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>transcript_id \"g1.t1\"; gene_id \"g1\";; gene_bio...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CAJQZA010000166.1</td>\n",
       "      <td>GeneMark.hmm3</td>\n",
       "      <td>exon</td>\n",
       "      <td>110</td>\n",
       "      <td>160</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>transcript_id \"g1.t1\"; gene_id \"g1\";; gene_bio...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>332331</th>\n",
       "      <td>OU015392.1</td>\n",
       "      <td>AUGUSTUS</td>\n",
       "      <td>transcript</td>\n",
       "      <td>108739437</td>\n",
       "      <td>108740144</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript_id \"g14364.t1\"; gene_id \"g14364\";; ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>332332</th>\n",
       "      <td>OU015392.1</td>\n",
       "      <td>AUGUSTUS</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>108739437</td>\n",
       "      <td>108739439</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>transcript_id \"g14364.t1\"; gene_id \"g14364\";; ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>332333</th>\n",
       "      <td>OU015392.1</td>\n",
       "      <td>AUGUSTUS</td>\n",
       "      <td>CDS</td>\n",
       "      <td>108739437</td>\n",
       "      <td>108740144</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>transcript_id \"g14364.t1\"; gene_id \"g14364\";; ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>332334</th>\n",
       "      <td>OU015392.1</td>\n",
       "      <td>AUGUSTUS</td>\n",
       "      <td>exon</td>\n",
       "      <td>108739437</td>\n",
       "      <td>108740144</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript_id \"g14364.t1\"; gene_id \"g14364\";; ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>332335</th>\n",
       "      <td>OU015392.1</td>\n",
       "      <td>AUGUSTUS</td>\n",
       "      <td>stop_codon</td>\n",
       "      <td>108740142</td>\n",
       "      <td>108740144</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>transcript_id \"g14364.t1\"; gene_id \"g14364\";; ...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>332336 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 scaffold    produced_by feature_type      start        end  \\\n",
       "0       CAJQZA010000166.1  GeneMark.hmm3         gene        110        817   \n",
       "1       CAJQZA010000166.1  GeneMark.hmm3   transcript        110        817   \n",
       "2       CAJQZA010000166.1  GeneMark.hmm3  start_codon        110        112   \n",
       "3       CAJQZA010000166.1  GeneMark.hmm3          CDS        110        160   \n",
       "4       CAJQZA010000166.1  GeneMark.hmm3         exon        110        160   \n",
       "...                   ...            ...          ...        ...        ...   \n",
       "332331         OU015392.1       AUGUSTUS   transcript  108739437  108740144   \n",
       "332332         OU015392.1       AUGUSTUS  start_codon  108739437  108739439   \n",
       "332333         OU015392.1       AUGUSTUS          CDS  108739437  108740144   \n",
       "332334         OU015392.1       AUGUSTUS         exon  108739437  108740144   \n",
       "332335         OU015392.1       AUGUSTUS   stop_codon  108740142  108740144   \n",
       "\n",
       "       dot strand frame                                               desc  \n",
       "0        .      +     .  transcript_id \"g1.t1\"; gene_id \"g1\";; gene_bio...  \n",
       "1        .      +     .  transcript_id \"g1.t1\"; gene_id \"g1\";; gene_bio...  \n",
       "2        .      +     0  transcript_id \"g1.t1\"; gene_id \"g1\";; gene_bio...  \n",
       "3        .      +     0  transcript_id \"g1.t1\"; gene_id \"g1\";; gene_bio...  \n",
       "4        .      +     0  transcript_id \"g1.t1\"; gene_id \"g1\";; gene_bio...  \n",
       "...     ..    ...   ...                                                ...  \n",
       "332331   1      +     .  transcript_id \"g14364.t1\"; gene_id \"g14364\";; ...  \n",
       "332332   .      +     0  transcript_id \"g14364.t1\"; gene_id \"g14364\";; ...  \n",
       "332333   1      +     0  transcript_id \"g14364.t1\"; gene_id \"g14364\";; ...  \n",
       "332334   .      +     .  transcript_id \"g14364.t1\"; gene_id \"g14364\";; ...  \n",
       "332335   .      +     0  transcript_id \"g14364.t1\"; gene_id \"g14364\";; ...  \n",
       "\n",
       "[332336 rows x 9 columns]"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "process_annotation(annotation_path).process_description()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "c6c96666",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['g1', 'g1.t1', 'transcript_id \"g1.t1\"; gene_id \"g1\";', ...,\n",
       "       'g14364', 'g14364.t1',\n",
       "       'transcript_id \"g14364.t1\"; gene_id \"g14364\";'],\n",
       "      shape=(46162,), dtype=object)"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "annotation.desc.unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "8f6a9a19",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['transcript_id \"g1.t1\"; gene_id \"g1\";',\n",
       "       'transcript_id \"g2.t1\"; gene_id \"g2\";',\n",
       "       'transcript_id \"g3.t1\"; gene_id \"g3\";', ...,\n",
       "       'transcript_id \"g14362.t1\"; gene_id \"g14362\";',\n",
       "       'transcript_id \"g14363.t1\"; gene_id \"g14363\";',\n",
       "       'transcript_id \"g14364.t1\"; gene_id \"g14364\";'],\n",
       "      shape=(15899,), dtype=object)"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "annotation.desc[annotation.desc.apply(lambda x: '_id' in x)].unique()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "rnaseq",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
